_posts/2020-04-11-native-mobile-to-flutter.markdown

```diff
@@ -0,0 +1,114 @@
+- Flutter offered clean, straight-forward conventions that were easy to learn, expressive and explicit.
+- Android Studio offered a turnkey development environment with performance monitoring tools, hot reload and step through debugging and inspection.  Since most of our developers are Android developers, the IDE was an added bonus due to its familiarity.
+- Flutter does not rely on any native UI components, so it's largely immune to breaking changes introduced by Apple, Google, or Android device manufacturers.
+
+### How much native code did we end up writing on both iOS and Android? 
+
+Flutter promises a very high percentage of shared cross-platform code across the platforms, but we still had to write some native code. 
+
+Flutter offers a very simple message passing API for communicating with native code.  We needed to use native integrations for the following relatively small group of features: 
+
+- Push notification.
+- Deep link handling.
+- Connectivity detection.
+- Security features.
+- Device information. 
+
+_This platform specific code amounted to less than 2% of our total lines of code._ 
+
+We were able to write this functionality easily and communicate between Flutter and native with very little interface overhead. Flutter provides a [very simple api](https://flutter.dev/docs/development/platform-integration/platform-channels#architecture) for communicating bidirectionally between native and dart code.
+
+### How was the learning curve?
+
+Dart/Flutter is definitely the easiest (of many) languages and UI frameworks I've ever learned.  Within two weeks most of our engineers were up to speed and highly productive. 
+
+### How secure is Flutter out of the box? 
+
+After completing our application, we hired an external security group to do an analysis on our application.  The analysis tests found no client side vulnerabilities.  The short answer is that **Flutter is secure out of the box**.  
+
+We didn't do that much to secure the app prior to the security test other than using an encrypted SQLite database and an excellent little plugin called [flutter_secure_storage](https://pub.dev/packages/flutter_secure_storage) which allowed us to write keys to the iOS Keychain or to Android Keystore.  We made sure that no data was persisted on the device in any location other than these two locations.  Aside from that, we mostly let the flutter and dart libraries do their thing. 
+
+All in all this was a pretty surprising result for us; we had initially expected to spend considerable time on a security deep dive after we were code complete.  So, imagine our surprise.
+
+### Challenges?
+
+We really wanted to add a challenges section to this article to give it a little balance.  Unfortunately, no one on the mobile team could identify anything that was particularly challenging.  We'll keep an eye out for gnarly bugs moving forward and will write follow ups when we finally encounter a good challenge to document.
```

Comment:
Maybe something along the lines of relatively immature 3rd-party packages?  Didn't we have to fork a package or two?

Reply:
- Yes - that's a good one - encrypted sqlite in particular.